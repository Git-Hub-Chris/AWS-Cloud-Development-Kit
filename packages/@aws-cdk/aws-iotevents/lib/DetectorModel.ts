import {
  IPrincipal,
  IRole,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "@aws-cdk/aws-iam";
import {
  Arn,
  Construct,
  IResource,
  Lazy,
  PhysicalName,
  Resource,
  Stack,
} from "@aws-cdk/core";
import { CfnDetectorModel } from "./iotevents.generated";
import { IState } from "./State";

/**
 * DetectorModel specification
 *
 * @export
 * @interface IDetectorModel
 * @extends {IResource}
 * @extends {IGrantable}
 */
export interface IDetectorModel extends IResource {
  /**
   * ARN for the {@link DetectorModel}
   *
   * @type {string}
   * @memberof IDetectorModel
   * @attribute
   */
  readonly detectorModelArn: string;
  /**
   * Name for the {@link DetectorModel}
   *
   * @type {string}
   * @memberof IDetectorModel
   * @attribute
   */
  readonly detectorModelName?: string;

  /**
   * The IAM service Role associated with this detector model.
   * This role is used as argument for {@link CfnDetectorModel} to
   * "grant permission to AWS IoT Events to perform its operations".
   *
   * @type {IRole}
   * @memberof IDetectorModel
   */
  readonly role?: IRole;
  /**
   * The grant principal for this Detector Model
   *
   * @type {IPrincipal}
   * @memberof IDetectorModel
   */
  readonly grantPrincipal?: IPrincipal;
  /**
   * Adds a statement to the IAM role assumed by the detector model
   *
   * @param {PolicyStatement} policyStatement
   * @memberof IDetectorModel
   */
  addToRolePolicy(policyStatement: PolicyStatement): void;
}

/**
 * Detector model attributes for import
 *
 * @export
 * @interface DetectorModelAttributes
 */
export interface DetectorModelAttributes {
  /**
   * Arn for the model
   *
   * @type {string}
   * @memberof DetectorModelAttributes
   */
  readonly detectorModelArn: string;

  /**
   * Name of the model
   *
   * @type {string}
   * @memberof DetectorModelAttributes
   * @default - generated from Arn
   */
  readonly detectorModelName?: string;

  /**
   * Arn for the role of the model
   *
   * @type {string}
   * @memberof DetectorModelAttributes
   * @default - no role
   */
  readonly roleArn?: string;
}
/**
 * Base class for an {@link DetectorModel}
 *
 * @export
 * @abstract
 * @class DetectorModelBase
 * @extends {Resource}
 * @implements {IDetectorModel}
 */
abstract class DetectorModelBase extends Resource
  implements IDetectorModel {
  public abstract readonly detectorModelArn: string;

  public abstract readonly detectorModelName?: string;
  public abstract readonly grantPrincipal?: IPrincipal;
  public abstract readonly role?: IRole;

  public addToRolePolicy(policyStatement: PolicyStatement): void {
    if (this.role) {
      this.role.addToPolicy(policyStatement);
    }
  }
}
/**
 * Props for Detector Model
 *
 * @export
 * @interface DetectorModelProps
 */
export interface DetectorModelProps {
  /**
   * Human readable description for the model
   *
   * @type {string}
   * @memberof DetectorModelProps
   * @default - no description
   */
  readonly description?: string;
  /**
   * Optional name for the model
   *
   * @type {string}
   * @memberof DetectorModelProps
   * @default - generated name
   */
  readonly detectorModelName?: string;
  /**
   * Service Role to assume while running the build.
   *
   * @default - A role will be created.
   */
  readonly role?: IRole;
  /**
   * The initial state
   *
   * @type {IState}
   * @memberof DetectorModelProps
   * @default - empty model
   */
  readonly entryPoint?: IState;

  /**
   * The key separating detectors
   *
   * @type {string}
   * @memberof DetectorModelProps
   * @default id
   */
  readonly key?: string;
}
/**
 * IoTEvents Detector Model
 *
 * @export
 * @class DetectorModel
 * @extends {DetectorModelBase}
 */
export class DetectorModel extends DetectorModelBase {
  public readonly detectorModelArn: string;
  public readonly detectorModelName?: string;
  public readonly grantPrincipal?: IPrincipal;
  public readonly role?: IRole;
  private initialState!: IState;
  public constructor(scope: Construct, id: string, props?: DetectorModelProps) {
    super(scope, id);
    this.role =
      (props && props.role) ||
      new Role(this, "Role", {
        roleName: PhysicalName.GENERATE_IF_NEEDED,
        assumedBy: new ServicePrincipal("iotevents.amazonaws.com"),
      });
    this.grantPrincipal = this.role;

    if (props && props.entryPoint) {
      this.initialState = props.entryPoint;
    }
    const resource = new CfnDetectorModel(this, "Resource", {
      detectorModelDescription: props && props.description,
      detectorModelName: props && props.detectorModelName,
      detectorModelDefinition: {
        initialStateName: Lazy.stringValue({
          produce: () =>
            this.initialState ? this.initialState.name : undefined,
        }),
        states: Lazy.anyValue({
          produce: () => {
            // Need a valid role to be able to process states and events
            if (this.role && this.initialState) {
              const visited: Set<IState> = new Set([this.initialState]);
              function visitor(state: IState) {
                state._relatedStates().forEach(nextState => {
                  if (!visited.has(nextState)) {
                    visited.add(nextState);
                    visitor(nextState);
                  }
                });
              }
              visitor(this.initialState);
              return [...visited].map(e => e._cfn(this.role!));
            } else {
              return [];
            }
          },
        }),
      },
      key: (props && props.key) || "id",
      roleArn: this.role ? this.role.roleArn : undefined,
    });
    this.detectorModelName = resource.detectorModelName || resource.ref;
    this.detectorModelArn = Arn.format(
      {
        service: "iotevents",
        resource: `detectorModel/${this.detectorModelName}`,
      },
      Stack.of(this)
    );
  }

  /**
   * Set the initial state for the state machine
   *
   * @param {IState} state
   * @memberof DetectorModel
   */
  public entryPoint(state: IState): void {
    this.initialState = state;
  }

  /**
   * Import detector model from Arn
   *
   * @param {Construct} scope
   * @param {string} id
   * @param {string} detectorModelArn
   * @returns {IDetectorModel}
   * @memberof DetectorModel
   */
  public fromDetectorModelArn(
    scope: Construct,
    id: string,
    detectorModelArn: string
  ): IDetectorModel {
    return this.fromDetectorModelAttributes(scope, id, { detectorModelArn });
  }

  /**
   * Import detector model
   *
   * @param {Construct} scope
   * @param {string} id
   * @param {DetectorModelAttributes} attrs
   * @returns {IDetectorModel}
   * @memberof DetectorModel
   */
  public fromDetectorModelAttributes(
    scope: Construct,
    id: string,
    attrs: DetectorModelAttributes
  ): IDetectorModel {
    const stack = Stack.of(scope);
    const detectorModelName =
      attrs.detectorModelName ||
      stack.parseArn(attrs.detectorModelArn).resource;

    const role = attrs.roleArn
      ? Role.fromRoleArn(stack, `${id}Role`, attrs.roleArn)
      : undefined;

    class Import extends DetectorModelBase {
      public readonly grantPrincipal = role;
      public readonly role = role;
      public readonly detectorModelArn = attrs.detectorModelArn;
      public readonly detectorModelName = detectorModelName;
    }

    return new Import(stack, id);
  }

  protected validate(): string[] {
    return [...this.validateInitialState()];
  }

  private validateInitialState(): string[] {
    if (!this.initialState) {
      return ["Missing initial state"];
    }
    return [];
  }
}
