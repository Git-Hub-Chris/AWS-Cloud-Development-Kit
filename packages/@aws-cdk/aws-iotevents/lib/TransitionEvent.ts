import { IGrantable } from "@aws-cdk/aws-iam";
import { AnyEventProperty, Event, IEvent } from "./Event";
import { IState } from "./State";
/**
 * Transition Event specification
 *
 * @export
 * @interface ITransitionEvent
 * @extends {IEvent}
 */
export interface ITransitionEvent extends IEvent {
  /**
   * The {@link IState} to transition into
   *
   * @type {IState}
   * @memberof ITransitionEvent
   */
  nextState: IState;
}

/**
 * IoTEvents Transition Event
 *
 * @export
 * @class TransitionEvent
 * @extends {Event}
 * @implements {ITransitionEvent}
 */
export class TransitionEvent extends Event implements ITransitionEvent {
  public nextState: IState;
  public constructor(name: string, nextState: IState, condition: string) {
    super(name, condition);
    this.nextState = nextState;
  }
  /**
   * @internal
   */
  public _cfn(grantable: IGrantable): AnyEventProperty {
    this.processGrants(grantable);
    return {
      eventName: this.name,
      condition: this.condition,
      actions: this.actions,
      nextState: this.nextState.name,
    };
  }
}
