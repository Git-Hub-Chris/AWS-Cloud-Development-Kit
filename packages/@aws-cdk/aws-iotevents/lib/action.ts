import * as iam from '@aws-cdk/aws-iam';
import { CfnDetectorModel } from './iotevents.generated';

/**
 * An abstract action for DetectorModel.
 */
export interface IAction {
  /**
   * The policies to perform the AWS IoT Events action.
   */
  readonly actionPolicies?: iam.PolicyStatement[];

  /**
   * Returns the AWS IoT Events action specification.
   */
  renderActionConfig(): ActionConfig;
}

/**
 * Properties for a AWS IoT Events action
 */
export interface ActionConfig {
  /**
   * The configuration for this action.
   */
  readonly configuration: CfnDetectorModel.ActionProperty;
}
