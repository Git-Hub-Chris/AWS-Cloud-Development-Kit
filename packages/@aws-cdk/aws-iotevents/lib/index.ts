export * from './input';
export * from './detector-model';
export * from './state';

// AWS::IoTEvents CloudFormation Resources:
export * from './iotevents.generated';
