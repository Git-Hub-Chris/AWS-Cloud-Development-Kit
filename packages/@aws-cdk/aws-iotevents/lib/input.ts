import * as iam from '@aws-cdk/aws-iam';
import { Resource, IResource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnInput } from './iotevents.generated';

/**
 * Represents an AWS IoT Events input
 */
export interface IInput extends IResource {
  /**
   * The name of the input
   * @attribute
   */
  readonly inputName: string;

  /**
   * The ARN of the input
   * @attribute
   */
  readonly inputArn: string;

  /**
   * Grant the indicated permissions on this input to the given IAM principal (Role/Group/User).
   *
   * @param grantee The principal (no-op if undefined)
   * @param actions The set of actions to allow (i.e. "dynamodb:PutItem", "dynamodb:GetItem", ...)
   */
  grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant

  /**
   * Grant the putting message permission to the given IAM principal (Role/Group/User).
   *
   * @param grantee The principal (no-op if undefined)
   */
  grantPutMessage(grantee: iam.IGrantable): iam.Grant
}


abstract class InputBase extends Resource implements IInput {
  /**
   * @attribute
   */
  public abstract readonly inputName: string;
  /**
   * @attribute
   */
  public abstract readonly inputArn: string;

  /**
   * Grant the indicated permissions on this input to the given IAM principal (Role/Group/User).
   *
   * @param grantee The principal (no-op if undefined)
   * @param actions The set of actions to allow (i.e. "dynamodb:PutItem", "dynamodb:GetItem", ...)
   */
  public grant(grantee: iam.IGrantable, ...actions: string[]) {
    return iam.Grant.addToPrincipal({
      grantee,
      actions,
      resourceArns: [this.inputArn],
      scope: this,
    });
  }

  /**
   * Grant the putting message permission to the given IAM principal (Role/Group/User).
   *
   * @param grantee The principal (no-op if undefined)
   */
  public grantPutMessage(grantee: iam.IGrantable) {
    return this.grant(grantee, 'iotevents:BatchPutMessage');
  }
}


/**
 * Properties for defining an AWS IoT Events input
 */
export interface InputProps {
  /**
   * The name of the input
   *
   * @default - CloudFormation will generate a unique name of the input
   */
  readonly inputName?: string,

  /**
   * An expression that specifies an attribute-value pair in a JSON structure.
   * Use this to specify an attribute from the JSON payload that is made available
   * by the input. Inputs are derived from messages sent to AWS IoT Events (BatchPutMessage).
   * Each such message contains a JSON payload. The attribute (and its paired value)
   * specified here are available for use in the condition expressions used by detectors.
   */
  readonly attributeJsonPaths: string[];
}

/**
 * Defines an AWS IoT Events input in this stack.
 */
export class Input extends InputBase {
  /**
   * Import an existing input
   */
  public static fromInputName(scope: Construct, id: string, inputName: string): IInput {
    return new class Import extends InputBase {
      public readonly inputName = inputName;
      public readonly inputArn = this.stack.formatArn({
        service: 'iotevents',
        resource: 'input',
        resourceName: inputName,
      });
    }(scope, id);
  }

  public readonly inputName: string;
  public readonly inputArn: string;

  constructor(scope: Construct, id: string, props: InputProps) {
    super(scope, id, {
      physicalName: props.inputName,
    });

    if (props.attributeJsonPaths.length === 0) {
      throw new Error('attributeJsonPaths property cannot be empty');
    }

    const resource = new CfnInput(this, 'Resource', {
      inputName: this.physicalName,
      inputDefinition: {
        attributes: props.attributeJsonPaths.map(path => ({ jsonPath: path })),
      },
    });

    this.inputName = this.getResourceNameAttribute(resource.ref);
    this.inputArn = this.stack.formatArn({
      service: 'iotevents',
      resource: 'input',
      resourceName: this.inputName,
    });
  }
}
