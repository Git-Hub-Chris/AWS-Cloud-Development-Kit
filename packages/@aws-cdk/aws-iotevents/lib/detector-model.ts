import * as iam from '@aws-cdk/aws-iam';
import { Resource } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnDetectorModel } from './iotevents.generated';
import { State } from './state';

/**
 * Properties for defining an AWS IoT Events detector model
 */
export interface DetectorModelProps {
  /**
   * The name of the detector model
   *
   * @default - CloudFormation will generate a unique name of the detector model
   */
  readonly detectorModelName?: string,

  /**
   * The state that is entered at the creation of each detector.
   */
  readonly initialState: State;
}

/**
 * Defines an AWS IoT Events detector model in this stack.
 */
export class DetectorModel extends Resource {
  constructor(scope: Construct, id: string, props: DetectorModelProps) {
    super(scope, id, {
      physicalName: props.detectorModelName,
    });

    const role = new iam.Role(this, 'DetectorModelRole', {
      assumedBy: new iam.ServicePrincipal('iotevents.amazonaws.com'),
    });

    new CfnDetectorModel(this, 'Resource', {
      detectorModelName: this.physicalName,
      detectorModelDefinition: {
        initialStateName: props.initialState.stateName,
        states: [props.initialState.toStateJson()],
      },
      roleArn: role.roleArn,
    });
  }
}
