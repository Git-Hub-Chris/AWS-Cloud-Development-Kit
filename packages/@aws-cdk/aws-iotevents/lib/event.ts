import { Expression } from './expression';
import { State } from './state';

/**
 * The base interface for events.
 */
interface IEventBase {
  /**
   * The name of the event.
   */
  readonly eventName: string;
}

/**
 * Specifies the actions to be performed when the condition evaluates to TRUE.
 */
export interface IEvent extends IEventBase {
  /**
   * The Boolean expression that, when TRUE, causes the actions to be performed.
   *
   * @default - none (the actions are always executed)
   */
  readonly condition?: Expression;
}

/**
 * Specifies the state transition and the actions to be performed when the condition evaluates to TRUE.
 */
export interface ITransitionEvent extends IEventBase {
  /**
   * The Boolean expression that, when TRUE, causes the state transition and the actions to be performed.
   */
  readonly condition: Expression;

  /**
   * The next state to transit to. When the resuld of condition expression is TRUE, the state is transited.
   */
  readonly nextState: State;
}
