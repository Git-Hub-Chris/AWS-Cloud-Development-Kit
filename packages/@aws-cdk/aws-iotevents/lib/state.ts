import { IEvent, ITransitionEvent } from './event';
import { CfnDetectorModel } from './iotevents.generated';

/**
 * Properties for defining a state of a detector
 */
export interface StateProps {
  /**
   * The name of the state.
   */
  readonly stateName: string;

  /**
   * Specifies the events on enter. the conditions of the events are evaluated when the state is entered.
   * If the condition is `TRUE`, the actions of the event are performed.
   *
   * @default - events on enter will not be set
   */
  readonly onEnter?: IEvent[];
}

/**
 * Defines a state of a detector
 */
export class State {
  /**
   * The name of the state
   */
  public readonly stateName: string;

  private transitionEvents: ITransitionEvent[] = []

  constructor(private readonly props: StateProps) {
    this.stateName = props.stateName;
  }

  /**
   * Add a transition event to the state.
   *
   * @param transitionEvent the transition event that triggered if condition is evaluated to TRUE
   */
  public transitionTo(transitionEvent:ITransitionEvent) {
    this.transitionEvents.push(transitionEvent);
  }

  /**
   * Return the JSON of the states that is transited to.
   * This function is called recursively and collect the states.
   *
   * @internal
   */
  public _getStatesJson(states: CfnDetectorModel.StateProperty[] = []): CfnDetectorModel.StateProperty[] {
    const { stateName, onEnter } = this.props;

    if (states.some(s => s.stateName === stateName)) {
      return states;
    }

    const newStates: CfnDetectorModel.StateProperty[] = [...states, {
      stateName,
      onInput: {
        transitionEvents: this.transitionEvents.length > 0 ?
          getTransitionEventJson(this.transitionEvents) : undefined,
      },
      onEnter: onEnter && { events: getEventJson(onEnter) },
    }];

    return this.transitionEvents.reduce((acc, transitionEvent) => {
      return transitionEvent.nextState._getStatesJson(acc);
    }, newStates);
  }

  /**
   * returns true if this state has at least one condition via events
   *
   * @internal
   */
  public _onEnterEventsHaveAtLeastOneCondition(): boolean {
    return this.props.onEnter?.some(event => event.condition) ?? false;
  }
}

function getEventJson(events: IEvent[]): CfnDetectorModel.EventProperty[] {
  return events.map(e => {
    return {
      eventName: e.eventName,
      condition: e.condition?.evaluate(),
    };
  });
}
function getTransitionEventJson(events: ITransitionEvent[]): CfnDetectorModel.TransitionEventProperty[] {
  return events.map(e => {
    return {
      eventName: e.eventName,
      condition: e.condition.evaluate(),
      nextState: e.nextState.stateName,
    };
  });
}
