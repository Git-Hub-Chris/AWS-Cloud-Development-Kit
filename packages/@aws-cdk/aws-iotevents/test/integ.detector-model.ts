import * as cdk from '@aws-cdk/core';
import * as iotevents from '../lib';

class TestStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const input = new iotevents.Input(this, 'MyInput', {
      inputName: 'test_input',
      attributeJsonPaths: ['payload.deviceId', 'payload.temperature'],
    });

    const firstState = new iotevents.State({
      stateName: 'firstState',
      onEnter: [{
        eventName: 'test-event',
        // meaning `condition: 'currentInput("test_input") && $input.test_input.payload.temperature == 31.5'`
        condition: iotevents.Expression.and(
          iotevents.Expression.currentInput(input),
          iotevents.Expression.eq(
            iotevents.Expression.inputAttribute(input, 'payload.temperature'),
            iotevents.Expression.fromString('31.5'),
          ),
        ),
      }],
    });
    const secondState = new iotevents.State({
      stateName: 'secondState',
    });

    // 1st => 2nd
    firstState.transitionTo({
      eventName: 'firstToSecond',
      nextState: secondState,
      condition: iotevents.Expression.eq(
        iotevents.Expression.inputAttribute(input, 'payload.temperature'),
        iotevents.Expression.fromString('12'),
      ),
    });
    // 2st => 1st
    secondState.transitionTo({
      eventName: 'secondToFirst',
      nextState: firstState,
      condition: iotevents.Expression.eq(
        iotevents.Expression.inputAttribute(input, 'payload.temperature'),
        iotevents.Expression.fromString('21'),
      ),
    });

    new iotevents.DetectorModel(this, 'MyDetectorModel', {
      detectorModelName: 'test-detector-model',
      description: 'test-detector-model-description',
      evaluationMethod: iotevents.EventEvaluation.SERIAL,
      detectorKey: 'payload.deviceId',
      initialState: firstState,
    });
  }
}

const app = new cdk.App();
new TestStack(app, 'detector-model-test-stack');
app.synth();
