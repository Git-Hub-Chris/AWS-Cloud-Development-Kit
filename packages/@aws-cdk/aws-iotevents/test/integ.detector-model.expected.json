{
  "Resources": {
    "MyInput08947B23": {
      "Type": "AWS::IoTEvents::Input",
      "Properties": {
        "InputDefinition": {
          "Attributes": [
            {
              "JsonPath": "payload.deviceId"
            },
            {
              "JsonPath": "payload.temperature"
            }
          ]
        },
        "InputName": "test_input"
      }
    },
    "MyDetectorModelDetectorModelRoleF2FB4D88": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "iotevents.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyDetectorModel559C0B0E": {
      "Type": "AWS::IoTEvents::DetectorModel",
      "Properties": {
        "DetectorModelDefinition": {
          "InitialStateName": "firstState",
          "States": [
            {
              "OnEnter": {
                "Events": [
                  {
                    "Condition": {
                      "Fn::Join": [
                        "",
                        [
                          "currentInput(\"",
                          {
                            "Ref": "MyInput08947B23"
                          },
                          "\") && $input.",
                          {
                            "Ref": "MyInput08947B23"
                          },
                          ".payload.temperature == 31.5"
                        ]
                      ]
                    },
                    "EventName": "test-event"
                  }
                ]
              },
              "OnInput": {
                "TransitionEvents": [
                  {
                    "Condition": {
                      "Fn::Join": [
                        "",
                        [
                          "$input.",
                          {
                            "Ref": "MyInput08947B23"
                          },
                          ".payload.temperature == 12"
                        ]
                      ]
                    },
                    "EventName": "firstToSecond",
                    "NextState": "secondState"
                  }
                ]
              },
              "StateName": "firstState"
            },
            {
              "OnInput": {
                "TransitionEvents": [
                  {
                    "Condition": {
                      "Fn::Join": [
                        "",
                        [
                          "$input.",
                          {
                            "Ref": "MyInput08947B23"
                          },
                          ".payload.temperature == 21"
                        ]
                      ]
                    },
                    "EventName": "secondToFirst",
                    "NextState": "firstState"
                  }
                ]
              },
              "StateName": "secondState"
            }
          ]
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "MyDetectorModelDetectorModelRoleF2FB4D88",
            "Arn"
          ]
        },
        "DetectorModelDescription": "test-detector-model-description",
        "DetectorModelName": "test-detector-model",
        "EvaluationMethod": "SERIAL",
        "Key": "payload.deviceId"
      }
    }
  }
}